サインアップ用のアクションを追加

a5ob7r opened the pull request:
# 概要

APIサーバにユーザ作成用のリクエストを送るためのアクションを追加する。

実際には、APIサーバへのリクエスト用のパラメータを送信させるフォームを送受信し、受け取ったパラメータを元に、APIサーバにユーザ作成用のリクエストを送信する。そのリクエストの結果を受けて、ユーザに対して適切なHTMLを返信する。

APIサーバにリクエストを送信する際に考慮している状況は以下の3つである。

- リクエストが成功する場合
- クライアントエラーによりリクエストが失敗する場合
- サーバエラーを含むその他の理由によりリクエストに失敗したりエラーが発生する場合

## 補足

~本PRは #4 の後続である。~

a5ob7r committed:
Add some gems to run system specs

a5ob7r committed:
Add the signup action

a5ob7r committed:
Add webmock gem

a5ob7r committed:
Add flash messages to each views

a5ob7r committed:
Add error messages to the user creation form

a5ob7r committed:
Enable aggregate_failures for rspec by default

a5ob7r committed:
Add rescue_from for StandardError

a5ob7r committed:
Add the user creation action

a5ob7r committed:
Add a empty action for login

a5ob7r committed:
Add the password confirmation for the user creation action

a5ob7r committed:
Add faker gem

a5ob7r committed:
Add flexirest gem

a5ob7r committed:
Add a missing environment variable for CI

a5ob7r committed:
Add a missing slash to the resource path

a5ob7r committed:
Add a missing environment variable to the rspec container

Taku3939 commented on compose.yaml:
```diff
@@ -13,6 +13,7 @@ services:
     ports:
       - "${APP_PORT:-3000}:3000"
     environment:
+      API_ENDPOINT_URL:
```
雑談: 環境変数同じ名前なら書かなくてもいいってことなんですかね？
知らなかった。

a5ob7r replied:
.envファイルから値を読み込んで設定する想定なのですが、このように読み込みたい環境変数を宣言しておかないと、コンテナ内で環境変数が設定されないと記憶しています。でも、経験上そうだったのでこのように書いているだけで、仕様がどうなっているかは自信がないですね。

Taku3939 replied:
僕が以前使っていた時は--env-fileで.env指定してcompose.yml内では`API_ENDPOINT_URL: ${API_ENDPOINT_URL}`と書いていたので、`API_ENDPOINT_URL:`だけで設定できるの知らなかったって話でした。(言葉足らずでした)

a5ob7r replied:
あー、そういう話でしたか。

ところで--env-fileオプションでわざわざ.envファイルを指定する必要ってあるんでしょうか。デフォルトで.envファイルを読んでくれると思っていました。単にcomposeファイルと.envファイルの配置されているディレクトリが異なるとかですかね。

Taku3939 replied:
別のenvファイルとスイッチして読み込みたかったか、その時には何かしらの理由があったはずなんですが、そこに関してはあまり覚えてないです。

tk0miya replied:
.env ファイルって勝手に読んでくれるんだっけ。あまり使わないので、そのあたりのお作法はよくわからない…

ただ、docker で環境を構築しようとしている場合は、環境ごとにこの値を変えるというのは特に意味を持たない気がするので、設定できないようにしても良さそうに思いました。

a5ob7r replied:
以前も同じような指摘をされましたね。ただ、今回はここにどういった値を設定するのがいいのか、今一つ確信が持てていないために空欄にしています。

ローカルでは、現状では `http://2023-api-training-server-app-1:3000` と設定していますが、これは、https://github.com/sls-training/2023-API-training-server の設定に依存するわけで、そのような設定を別のリポジトリである本リポジトリの設定ファイルに埋め込んでよいものか、と考えていました。

tk0miya replied:
そういえば、その部分ってどう動くんでしたっけ。普通に起動してしまうとそれぞれ独立した docker network になってしまい、通信できなさそうな気がするけど、うまく動くんです?

tk0miya replied:
某案件では API サーバの docker イメージを作って、それを client 側の docker-compose で起動するようにしています (作っている最中です)。

ある程度 API が成熟しているので、同時に開発することはないだろうという前提はありますね。

a5ob7r replied:
> そういえば、その部分ってどう動くんでしたっけ。普通に起動してしまうとそれぞれ独立した docker network になってしまい、通信できなさそうな気がするけど、うまく動くんです?

接続したいなら、手動で `docker network connect` でやるという方針です。 と言いつつ、できろうだろうと思っているだけで、本当にできるかは試していないです...。 `compose.yaml` に記述することもやってみましたが、前述のように他のリポジトリに依存する設定を埋め込むのはいかがなものか、と考えて不採用にしました。それにwebmock gemのおかげで、少なくともテストでは他のリポジトリのコンテナを利用しなくてもよいので、わざわざ `compose.yaml` に記述しなくてもいいかな、と考えています。

頻繁にAPIサーバと接続する必要が出てきたら、考えが変わるかもしれません。

tk0miya replied:
その方針ならたしかに compose.yaml に書くのは不適当ですね。納得したので取り下げます。

a5ob7r replied:
今気がつきましたが、 `API_ENDPOINT_URL: ${API_ENDPOINT_URL:-http://example.com:3000}` と書くと、デフォルト値を設定しつつ.envから値を任意に設定できますね(多分)。次書くときはこちらを採用しようかな。どんな値を設定することを想定しているのかを示すドキュメントとしての役割も期待できるし。

Taku3939 commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,12 @@
+
+  private
+
+  def render_internal_server_error
+    render file: Rails.root.join('public/500.html'), status: :internal_server_error,
+           layout: false, content_type: 'text/html'
+  end
```
雑談: 開発中にエラーみたかったらデバッガを使うかコメントアウトするしかないですか？(少し不便そうだと思った)

a5ob7r replied:
あー、そういうことは考えなかったですね。

a5ob7r replied:
ちなみに @Taku3939 だったらどうすると思いますか。

Taku3939 replied:
僕は`if Rails.env == 'production'`とかで開発環境ではログが表示されるようにするくらいしか思いつかないですね

a5ob7r replied:
> Using rescue_from with Exception or StandardError would cause serious side-effects as it prevents Rails from handling exceptions properly. As such, it is not recommended to do so unless there is a strong reason.

https://guides.rubyonrails.org/action_controller_overview.html#rescue-from

なるほど。安易にStandardErrorをrescue_fromで補足するべきではないようだ。実際、 @Taku3939 に指摘されたような影響があるわけで。

a5ob7r replied:
rescue_fromによるStandardErrorの補足を削除して、アクション内で必要な例外のみを捕捉するように変更しました。

Taku3939 commented on spec/system/users_spec.rb:
```diff
@@ -0,0 +1,133 @@
+    end
+  end
+
+  describe 'パスワードがパスワードの確認と一致しないとき' do
```
`〜の場合`を書くのだとcontextっぽい気がしました。
あとcontextが対照になっていないのも気になりました。
(システムスペックの場合でもリクエストと同じように、対照的に書くものなのか自分はわかってないです)

a5ob7r replied:
> `〜の場合`を書くのだとcontextっぽい気がしました。

おっと、ミスですね。修正します。

> あとcontextが対照になっていないのも気になりました。
> (システムスペックの場合でもリクエストと同じように、対照的に書くものなのか自分はわかってないです)

対照というと、例えばどのようになっていればよいのでしょうか。

Taku3939 replied:
```ruby
context "パスワードが間違っている場合" do
 it "ログインできずエラーメッセージを表示する" do
 end
end

context "パスワードが正しい場合" do
 it "ログインできてリダイレクトする" do
 end
end
```
↑のように対照的に書くものなのかなと思っていました。

a5ob7r replied:
今回の場合だと

```ruby
context 'パスワードがパスワードの確認と一致しないとき'
context 'パスワードがパスワードの確認と一致するとき'
```

とするべきということでしょうか。

Taku3939 replied:
そうですね。僕はそのほうが良いと思います。

a5ob7r replied:
なるほど。ですが、今回の場合であれば、パスワードがパスワードの確認と一致していたとしても、他のパラメータも正当でないと、結局リクエストに失敗するわけですよね。他のパラメータも正当であるなら、それは全てのパラメータが正当である場合のスペックがありますし、正当でない場合のスペックも書くことは可能ですが、結局リクエストに失敗するというのであれば、パスワードがパスワードの確認と一致しない場合のみで現状は十分ではないでしょうか。

まあ、パスワード一致の確認さえ満たせば、他のパラメータが不正でもリクエストに成功してしまう、みたいな場合がないかを確認したいとかなら、書くのもありか。

Taku3939 replied:
まあ確かに、そうかもです。

けど今のままだと以下の点で読み取りづらい気はしました(素人考えで全然あってるかはわからないですが自分の考えだけ貼っておきます)
- システムスペックであることを考えると`フォームの入力`な気がする。
- contextの内容が対照的に書いても良いと思うこと

僕が考えてるイメージですが、、まちがってたらすみません
```ruby
describe "ユーザの作成" do
 context 'フォームの入力が正当であるとき' do
 　 it  'ログイン用のURLにリダイレクトする' 
 　 it 'ユーザの作成に成功した旨を示すメッセージが表示される'
 end

 context 'フォームの入力が不正であるとき' do
   context 'passwordがパスワードの確認と一致しないとき' do
     it 'サインアップ用のページを表示する'
     it 'パラメータに関するバリデーションエラーが表示される'
   end
   context 'パスワードがパスワードの確認と一致しているとき' do 
     it 'サインアップ用のページを表示する'
     it  'パスワードがパスワードの確認と一致しない旨のバリデーションエラーが表示される' 
   end
end
```

a5ob7r replied:
あー、確かに文言についてはもう少しやりようがあるかな。修正します。その過程で、スペックの構成も考えてみます。

a5ob7r replied:
スペックの構成を整理してみました。

```ruby
RSpec.describe 'User Creation', type: :system do
  context 'フォームへの入力が全て正当であるとき' do
    context 'APIサーバから成功レスポンスが返ってきたとき'
    context 'APIサーバからエラーレスポンスが返ってきたとき'
  context 'フォームへの入力に不正があるとき'
    context '各フォームへの入力が空であるとき'
    context 'パスワードがパスワードの確認と一致しないとき'
  end
end
```

今回はやはりスペックを対称にする必要はないと考えていますが、一方でフォームへの入力という観点で考えると、スペックの構成が整理されてスッキリしたように思います。ありがとうございました。

Taku3939 commented on compose.yaml:
```diff
@@ -13,6 +13,7 @@ services:
     ports:
       - "${APP_PORT:-3000}:3000"
     environment:
+      API_ENDPOINT_URL:
```
ちょっと試しにテストで

a5ob7r committed:
Fix incorrect example method names

a5ob7r committed:
Change the spec file name

a5ob7r committed:
Sort the user creation spec for understandability

a5ob7r committed:
Add a rescue statement for Flexirest's some exceptions

a5ob7r committed:
Remove a global rescue_from statement for StandardError

Rails Guide doesn't recommended it.
https://guides.rubyonrails.org/action_controller_overview.html#rescue-from

tk0miya commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,34 @@
+
+  def confirm_password
+    return if params.dig(:user, :password).blank? ||
+      params[:user][:password] == params.dig(:user, :password_confirmation)
```
なんで左辺は `[]` で右辺は `#dig` なんですっけ。両方 `#dig` で良い気がする

a5ob7r replied:
強いていうなら、趣味といったところでしょうか。左辺がfalseになっているなら、右辺では `params[:user][:password]` のようにアクセスしても例外が発生しないような値の構造になっていることがわかるので、このようになっています。元々#digを使わないで書いていた部分がそのままになっているということでもありますが。

ここでは、どちらでも処理上は同じ結果になると思いますが、揃っているほうが、前述のようなことを考えなくて良い分、読み手の認知負荷が低くていいんですかね。

tk0miya replied:
ひとつ上の 28行目では dig を使っていますよね。
質問では左辺と右辺という書き方をしたのですが、if 文全体としては
```
if dig || [] == dig
```
という構成ですよね。これ全体で趣味ということであっていますか?

>ここでは、どちらでも処理上は同じ結果になると思いますが、揃っているほうが、前述のようなことを考えなくて良い分、読み手の認知負荷が低くていいんですかね。

多段の Hash アクセスはエラーの元なので、個人的には特に意図がなければなるべく避けたいというのがありますね。
今回の場合は事前に dig をしているので安全であるというのを、レビュー段階では確認していました。

※ rubocop のルールになってないみたいですね。提案したほうがいいぐらいのやつだと思ってます。

a5ob7r replied:
> ひとつ上の 28行目では dig を使っていますよね。 質問では左辺と右辺という書き方をしたのですが、if 文全体としては
> 
> ```
> if dig || [] == dig
> ```
> 
> という構成ですよね。これ全体で趣味ということであっていますか?
> 

あー、 `dig || 左辺 == 右辺` ではなく、 `左辺 || 右辺` と解釈していました。とはいえ、これ全体が趣味ということ自体は合っています。

> > ここでは、どちらでも処理上は同じ結果になると思いますが、揃っているほうが、前述のようなことを考えなくて良い分、読み手の認知負荷が低くていいんですかね。
> 
> 多段の Hash アクセスはエラーの元なので、個人的には特に意図がなければなるべく避けたいというのがありますね。 今回の場合は事前に dig をしているので安全であるというのを、レビュー段階では確認していました。
> 
> ※ rubocop のルールになってないみたいですね。提案したほうがいいぐらいのやつだと思ってます。

意図があるかないかで言えば前述の通りあるのですが、それによるリターンが特にないので、統一してしまっていいのかなと思い始めています。

tk0miya replied:
dig した後に同じデータに [] でアクセスするのはなにか意図があるのではと勘ぐってしまいそうで、僕は認知負荷が高まるだろうなと思います。
敢えて違う書き方をしているのだから意味があるはずだろう、と僕は読みがちですね

tk0miya commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  rescue Flexirest::HTTPClientException => e
+    @errors = e.result.errors
+    render 'new', status: :unprocessable_entity
+  rescue Flexirest::HTTPServerException, Flexirest::TimeoutException, Flexirest::ConnectionFailedException
+    render_internal_server_error
```
このエラーハンドリングに気づけてるのはナイスですね。
API コールするときは、接続やタイムアウトといった問題が起こりえます。レスポンスが帰ってくるケースばかりに気を取られず、こうしたエラーが帰ってくることも考慮に入れて実装できるといいですね。

※ 今回は internal server error にしていますが、どうエラーハンドリングするのかはアプリ次第です。単に再 render するという選択であっても、この動きでもどちらでもよさそうです。

a5ob7r replied:
> 単に再 render するという選択であっても

なるほど、そういう選択肢もあるんですね。思い至りませんでした。むしろそちらのほうがいいかな。

tk0miya replied:
なぜ API サーバとの通信がうまくいかなかったのかにも寄るのですが、トラブルによっては復旧に時間がかかることもありうるので、単に再 render でよいのかはシステム次第 / インフラ次第ですね。
単に瞬断が起きていただけなのか (AWS 自体のトラブルでじきに復旧するのか)、自動復旧の仕組みを備えていて待てば解決するのか、人間が対処するまで直らないのか、というのはエラーコードでは識別できないですしね。

単に再 render するだけだと、そのエラーの深刻さがユーザーに伝わりづらいので、どういうメッセージを出すかちょっと悩むところです。
しばらく経ってから再試行してください、的なメッセージを出したい気もしますね。

tk0miya commented on config/routes.rb:
```diff
@@ -1,6 +1,6 @@
 Rails.application.routes.draw do
-  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
+  get 'signup', to: 'users#new'
+  get 'login', to: 'sessions#new'
```
こうやって読み替えを行うのなら UserController と名付けるのではなく、SignupController と名付けたほうがいいんじゃないかなあ。なぜ読み替えをここで行う必要があるのかよくわかりませんでした。

なぜ入力フォームは /signup で、POST 先は /users なのか、敢えて複雑にしているように見えます。

a5ob7r replied:
以前に同様の指摘をされた箇所ですね。Railsチュートリアルでこのような命名をしていたので、脳死でそれに倣っていました。ですが、確かに必要性は感じないので通常のルーティングに従うことにします。

やるにしても、 `users#new` への代理リソースとして設定しておくくらいですかね。

tk0miya commented on spec/system/user_creation_spec.rb:
```diff
@@ -0,0 +1,118 @@
+        WebMock.stub_request(:post, File.join(Api::User.base_url, 'user')).to_return status: 500
+      end
+
+      it '500エラー用のページを表示する' do
```
500エラー用といいつつ、200 OK が返っていますよね。it 節が微妙に違うかも。

a5ob7r replied:
あれ、そのようになっていますかね。指摘の意図がよくわかりませんでした。もしかして、ステータスコードを確認するスペックがないので、そのように読み取れるということでしょうか。

tk0miya replied:
あ、読み間違えてました。指摘間違いです。ごめんなさい。

tk0miya commented on spec/system/user_creation_spec.rb:
```diff
@@ -0,0 +1,118 @@
+    context 'パスワードがパスワードの確認と一致しないとき' do
+      let(:password_confirmation) { Faker::Internet.password }
+
+      before do
+        WebMock.stub_request(:post, File.join(Api::User.base_url, 'user')).to_return status: 200
+      end
```
この stub って必要なんでしたっけ?

webmock や allow の厄介なところは、間違った設定とバグったコードの組み合わせがうまく動くように見えるところです。通信をしていなくてもこれが通ってしまうところです。
ですから、呼び出されたことを it 節でちゃんと確認したいところですね。

a5ob7r replied:
> この stub って必要なんでしたっけ?

一応、なくても動作します。ただ、このstubの意図としては、ここでのパスワード確認のロジックが機能しないと、送信されたパスワードとパスワード確認が一致していないにも関わらず、リクエストが成功してしまうという状況を安全確実に作りたい、となります。このstubがないと、実際にAPIサーバにリクエストが送信され、ユーザが作成されてしまいます。おそらく、外部のDBに対応するレコードが作成されるわけですが、外部のDBですから、通常はテスト終了後のクリーニング処理も適用されないでしょうから、あまり望ましいことであるとは思えませんでした。

ちなみに、パスワードの確認は本クライアントサーバ上でのみ行うもので、APIサーバでは確認しません。というか、確認用パスワードのパラメータをサポートしていません。

tk0miya replied:
webmock を有効にしている場合、stub を設定していない場所とは通信できないようになっているはずですよ。
指定した箇所の通信だけを塞ぐのではなく、全部塞ぐというポリシーだったはずです。

ですから

>ただ、このstubの意図としては、ここでのパスワード確認のロジックが機能しないと、送信されたパスワードとパスワード確認が一致していないにも関わらず、リクエストが成功してしまうという状況を安全確実に作りたい、となります。

というのは勘違いじゃないかと思いました。

a5ob7r replied:
そうなんですね。ちゃんとドキュメントを読んでいませんでした。確認して出直してきます。

a5ob7r replied:
READMEには、そのものズバリな記述は見つけられなかったけど、そうなのかな、と思える記述はあった。

> WebMock API was also designed with connect being part of request step, and it only allows stubbing requests, not connections.

https://github.com/bblimke/webmock#connecting-on-nethttpstart

それに、実際にスタブしていないリクエストを投げようとしてみると、webmockがHTTP接続を許可していない旨のエラーが発生した。

a5ob7r replied:
この挙動って、webmock系のライブラリだとデファクトなんですかね。

tk0miya replied:
>この挙動って、webmock系のライブラリだとデファクトなんですかね。

長らく webmock しかつかっていないので、他のライブラリでどうなっているかは自信がないですね。

a5ob7r committed:
Add `/users/new` to sign up instead of `/signup`

No any good reasons to use this route naming.

a5ob7r committed:
Remove an unnecessary single `subject` call

tim-toga commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  def confirm_password
+    return if params.dig(:user, :password).blank? ||
```
passwordがblankな場合はtrueで合ってます？

tim-toga replied:
blankはモデルで検証したいからか

a5ob7r replied:
意図としてはそうなります。ですが、ややこしいので、値に関係なくパスワードとパスワードの確認が一致するかを確認するだけでいいかな、と思い始めています。

tim-toga replied:
良いと思います

a5ob7r committed:
Change the password confirmation check condition

Check it even if the password value is nil.

a5ob7r committed:
Remove an unnecessary stub

By default webmock gem disallows not-stubed requests.

a5ob7r commented:
指摘された箇所は一通り修正しました。

a5ob7r committed:
Change the routing name for the login action

No any good reasons to use another name for me.

a5ob7r commented:
ありがとうございました。
